test/clients/openshift/client.go

```diff
@@ -67,80 +72,209 @@ func newClientFromKubeConfig(kc *api.Config) (*Client, error) {
-func NewAzureClusterReaderClient() (*Client, error) {
+func newAzureClusterReaderClient(cs *internalapi.OpenShiftManagedCluster) (*client, error) {
 	if os.Getenv("KUBECONFIG") != "" {
 		return clientFromEnv()
 	}
-	dataDir, err := fakerp.FindDirectory(fakerp.DataDirectory)
-	if err != nil {
-		return nil, err
-	}
-	cs, err := managedcluster.ReadConfig(filepath.Join(dataDir, "containerservice.yaml"))
-	if err != nil {
-		return nil, err
-	}
 
 	var kc api.Config
-	err = latest.Scheme.Convert(cs.Config.AzureClusterReaderKubeconfig, &kc, nil)
+	err := latest.Scheme.Convert(cs.Config.AzureClusterReaderKubeconfig, &kc, nil)
 	if err != nil {
 		return nil, err
 	}
 
 	return newClientFromKubeConfig(&kc)
 }
 
-func NewCustomerReaderClient() (*Client, error) {
+func newCustomerReaderClient(cs *internalapi.OpenShiftManagedCluster) (*client, error) {
 	if os.Getenv("KUBECONFIG") != "" {
 		return clientFromEnv()
 	}
-	kc, err := login("customer-cluster-reader")
+	kc, err := login("customer-cluster-reader", cs)
 	if err != nil {
 		return nil, err
 	}
 
 	return newClientFromKubeConfig(kc)
 }
 
-func NewAdminClient() (*Client, error) {
+func newAdminClient(cs *internalapi.OpenShiftManagedCluster) (*client, error) {
 	if os.Getenv("KUBECONFIG") != "" {
 		return clientFromEnv()
 	}
-	kc, err := login("admin")
+	kc, err := login("admin", cs)
 	if err != nil {
 		return nil, err
 	}
 
 	return newClientFromKubeConfig(kc)
 }
 
-func NewCustomerAdminClient() (*Client, error) {
+func newCustomerAdminClient(cs *internalapi.OpenShiftManagedCluster) (*client, error) {
 	if os.Getenv("KUBECONFIG") != "" {
 		return clientFromEnv()
 	}
-	kc, err := login("customer-cluster-admin")
+	kc, err := login("customer-cluster-admin", cs)
 	if err != nil {
 		return nil, err
 	}
 
 	return newClientFromKubeConfig(kc)
 }
 
-func NewEndUserClient() (*Client, error) {
+func newEndUserClient(cs *internalapi.OpenShiftManagedCluster) (*client, error) {
 	if os.Getenv("KUBECONFIG") != "" {
 		return clientFromEnv()
 	}
-	kc, err := login("enduser")
+	kc, err := login("enduser", cs)
 	if err != nil {
 		return nil, err
 	}
 
 	return newClientFromKubeConfig(kc)
 }
+
+type Client struct {
+	log                *logrus.Entry
+	cs                 *internalapi.OpenShiftManagedCluster
+	AzureClusterReader *client
+	CustomerReader     *client
+	Admin              *client
+	CustomerAdmin      *client
+	EndUser            *client
+}
+
+var _ clients.DeepTestInterface = &Client{}
+
+// NewClient creates a new deep testing client for OpenshiftManagedCluster resources.
+// If the E2E tests are run against the fake rp then parameters should be left nil and
+// sensible defaults for the logger and cluster struct will be set.
+// If the E2E tests are run against the real rp then the logger and/or the cluster config
+// should be set.
+func NewClient(log *logrus.Entry, cs *internalapi.OpenShiftManagedCluster) *Client {
+	c := &Client{
+		log: log,
+		cs:  cs,
+	}
+	var err error
+	if c.log == nil {
```

Comment:
FWIW I feel that the implicit pattern of defaulting nil always goes sour in the long run, especially in test libraries.  I personally prefer explicit even if it's verbose.  Even `func NewDefaultClient() *Client { return NewClient(testlogging.GetTestLogger(), c.discoverInternalConfig() }` or whatever is better.